proposals/gc/MVP.md

```diff
@@ -0,0 +1,272 @@
+# GC v1 Extensions
+
+
+## Language
+
+Based on reference types proposal.
+
+### Types
+
+#### Value Types
+
+* `eqref` is a new reference type
+  - `reftype ::= ... | eqref`
+
+* `ref <typeidx>` is a new reference type
+  - `reftype ::= ... | ref <typeidx>`
+  - `ref $t ok` iff `$t` is defined in the context
+
+* `int31ref` is a new reference type
+  - `reftype ::= ... | int31ref`
+
+
+#### Type Definitions
+
+* `deftype` is the new category of types that can occur as type definitions
+  - `deftype ::= <functype> | <structtype> | <arraytype>`
+  - `module ::= {..., types vec(<deftype>)}`
+
+* `structtype` describes a structure with statically indexed fields
+  - `structtype ::= struct <fieldtype>*`
+
+* `arraytype` describes an array with dynamically indexed fields
+  - `arraytype ::= array <fieldtype>`
+
+* `fieldtype` describes a struct or array field and whether it is mutable
+  - `fieldtype ::= <mutability> <storagetype>`
+  - `storagetype ::= <valtype> | <packedtype>`
+  - `packedtype ::= i8 | i16`
+
+* Unpacking a storage type yields `i32` for packed types, otherwise the type itself
+  - `unpacked(t) = t`
+  - `unpacked(pt) = i32`
+
+
+#### Imports
+
+* `type <typetype>` is an import description with an upper bound
+  - `importdesc ::= ... | type <reftype>`
+  - Note: `type` may get additional parameters in the future
```

Comment:
Could you explain the reasoning behind having the type have a `reftype` as opposed to a `deftype`?  E.g., into which index space does a type import belong?  It seems like it should be in the type definition index space (so it can be used in function types, exports, etc).  I was imagining that maybe, as a special case (to play well with section ordering), type imports would go into the type section (using a special import `form` followed by the type constructor's `form`).  Then a type import would be prepended to both the type definition index space and the import index space.  Kinda weird, but seems forced by the type/import circularity.

Reply:
Well, yes, except that type imports are still meant to be prepended to the type index space. (The ordering of binary sections doesn't have to imply the ordering of index spaces. As you said, they need to be validated together anyway.)

Added comments.

Reply:
Regarding the bound: In most cases, it will just be `anyref`, i.e., the type is fully abstract. How would you express that if it wasn't a reftype?

FWIW, I'm not sure I understood what you meant by:
> In particular, it seems like the import needs to introduce a new type definition; it can't just reference an existing one.

Why not?

Reply:
Ok, I think I understand the proposal a bit better now.  I also see the benefit in that, by using a reference-type bound (instead of using a struct/array bound more-directly), one has the more-expressive capability of declaring whether the import is equality-comparable, nullable, etc.